packages/alfa-rules/src/sia-r59/rule.ts

```diff
@@ -0,0 +1,50 @@
+import { Rule } from "@siteimprove/alfa-act";
+import { Document, Element, Namespace } from "@siteimprove/alfa-dom";
+import { Iterable } from "@siteimprove/alfa-iterable";
+import { Predicate } from "@siteimprove/alfa-predicate";
+import { Err, Ok } from "@siteimprove/alfa-result";
+import { Page } from "@siteimprove/alfa-web";
+
+import { hasChild } from "../common/predicate/has-child";
+import { hasName } from "../common/predicate/has-name";
+import { hasNamespace } from "../common/predicate/has-namespace";
+import { hasRole } from "../common/predicate/has-role";
+import { isDocumentElement } from "../common/predicate/is-document-element";
+
+const { filter, isEmpty } = Iterable;
+const { and, equals, test } = Predicate;
+
+export default Rule.Atomic.of<Page, Document>({
+  uri: "https://siteimprove.github.io/sanshikan/rules/sia-r59.html",
+  evaluate({ document }) {
+    return {
+      applicability() {
+        return test(
+          hasChild(and(Element.isElement, isDocumentElement())),
+          document
+        )
+          ? [document]
+          : [];
+      },
+
+      expectations(target) {
+        const headings = filter(
+          target.descendants({ flattened: true }),
+          and(
+            Element.isElement,
+            and(
+              hasNamespace(equals(Namespace.HTML)),
+              hasRole(hasName(equals("heading")))
+            )
+          )
+        );
+
+        return {
+          1: isEmpty(headings)
+            ? Err.of("The document does not have a heading element")
+            : Ok.of("The document has at least one heading element")
```

Comment:
Rather than do `isEmpty(filter(..))` you can simply do `some(...)` and then switch around `Err` and `Ok`.

Reply:
🤔 but `some` is expecting a predicate to test on the iterable and has no default predicate.
This means I have to do `some(heading, isElement)` or something similar which is actually less readable I think (and feels a bit redundant).

=> Should `some` have a default predicate (always true) for this use case?

Reply:
To clarify, you'd do this:

```ts
const hasHeadings = some(
  target.descendants({ flattened: true }),
  and(
    Element.isElement,
    and(
      hasNamespace(equals(Namespace.HTML)),
      hasRole(hasName(equals("heading")))
    )
  )
)
```

Reply:
🤦‍♂ 
Of course…